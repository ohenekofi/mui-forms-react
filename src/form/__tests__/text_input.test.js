import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { FormProvider, useForm } from "react-hook-form";
import TextInput from "../text_input";

const errorMessage = "This is the error message";
const helperText = "This is the helper text.";
const label = "Test Label";
const maxLength = 127;
const maxLengthForError = 7;
const name = "testName";
const value = "This is what the user types in.";

const Wrapped = (props) => {
  const methods = useForm({
    defaultValues: {
      [name]: "",
    },
  });
  return (
    <FormProvider {...methods}>
      <TextInput name={name} label={label} {...props} />
    </FormProvider>
  );
};

describe("<TextInput />", () => {
  it("renders without crashing", () => {
    render(<Wrapped />);
    expect(screen.getByRole("textbox")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { container } = render(<Wrapped />);
    expect(container).toMatchSnapshot();
  });

  it("displays the label", () => {
    render(<Wrapped />);
    expect(screen.getByText(label)).toBeInTheDocument();
  });

  it("displays the value", () => {
    render(<Wrapped />);
    fireEvent.change(screen.getByRole("textbox"), {
      target: { value },
    });
    expect(screen.getByText(value)).toBeInTheDocument();
  });

  it("displays the helper text", () => {
    render(<Wrapped helperText={helperText} />);
    expect(screen.getByText(helperText)).toBeInTheDocument();
  });

  it("displays the required indicator", () => {
    render(<Wrapped rules={{ required: true }} />);
    expect(screen.getByText("*")).toBeInTheDocument();
  });

  it("displays the max length indicator", () => {
    render(
      <Wrapped
        rules={{
          maxLength: {
            value: maxLength,
            message: `Name must be less than ${maxLength} characters.`,
          },
        }}
      />
    );
    expect(screen.getByText(`0/${maxLength}`)).toBeInTheDocument();
  });

  it("displays the error message", async () => {
    const user = userEvent.setup();
    render(
      <Wrapped
        rules={{
          maxLength: {
            value: maxLengthForError,
            message: errorMessage,
          },
        }}
      />
    );
    await user.click(screen.getByRole("textbox"));
    await user.keyboard(value);
    expect(screen.getByText(value)).toBeInTheDocument();
    expect(
      screen.getByText(`${value.length}/${maxLengthForError}`)
    ).toBeInTheDocument();
    await waitFor(() => {
      expect(screen.getByText(errorMessage)).toBeInTheDocument();
    });
  });
});
