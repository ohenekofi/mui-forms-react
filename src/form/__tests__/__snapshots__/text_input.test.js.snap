// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInput /> matches snapshot 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-10"
  >
    <span
      class="MuiTypography-root makeStyles-caption-6 MuiTypography-caption"
    >
      <div>
        Â 
      </div>
    </span>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiFormLabel-colorSecondary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink="false"
        for="text-input-testName"
      >
        Test Label
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorSecondary MuiInput-colorSecondary MuiInputBase-formControl MuiInput-formControl MuiInputBase-multiline MuiInput-multiline MuiInputBase-adornedEnd"
      >
        <textarea
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline MuiInput-inputMultiline MuiInputBase-inputAdornedEnd"
          id="text-input-testName"
          placeholder=""
          rows="1"
          style="height: 0px; overflow: hidden;"
        />
        <textarea
          aria-hidden="true"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline MuiInput-inputMultiline MuiInputBase-inputAdornedEnd"
          readonly=""
          style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); width: 100%;"
          tabindex="-1"
        />
      </div>
      <p
        class="MuiFormHelperText-root makeStyles-helperText-8"
        role="note"
      />
    </div>
  </div>
</div>
`;
